File: frontend/lib/api/keywords.ts
/**
 * キーワード関連のAPIクライアント関数
 */
import { get, post } from './client';
import type { Keyword } from '@/types/knowledge';

/**
 * キーワード一覧を取得
 * @param limit 取得件数の上限（デフォルト: 100）
 * @returns キーワード一覧
 */
export async function getKeywords(limit = 100): Promise<Keyword[]> {
  const queryParams = new URLSearchParams();
  if (limit) {
    queryParams.append('limit', String(limit));
  }

  const queryString = queryParams.toString();
  const endpoint = `/api/keywords${queryString ? `?${queryString}` : ''}`;
  
  return get<Keyword[]>(endpoint);
}

/**
 * キーワード検索
 * @param query 検索クエリ
 * @param limit 取得件数の上限（デフォルト: 20）
 * @returns 検索結果のキーワード一覧
 */
export async function searchKeywords(
  query: string,
  limit = 20
): Promise<Keyword[]> {
  const queryParams = new URLSearchParams();
  queryParams.append('q', query);
  if (limit) {
    queryParams.append('limit', String(limit));
  }

  const endpoint = `/api/keywords/search?${queryParams.toString()}`;
  return get<Keyword[]>(endpoint);
}

/**
 * キーワードを作成（存在しない場合は作成、存在する場合は既存のものを返す）
 * @param name キーワード名
 * @returns 作成されたまたは既存のキーワード
 */
export async function createKeyword(name: string): Promise<Keyword> {
  return post<Keyword>('/api/keywords', { name });
}

/**
 * 人気のキーワードを取得（使用回数順）
 * @param limit 取得件数の上限（デフォルト: 10）
 * @returns 人気のキーワード一覧
 */
export async function getPopularKeywords(limit = 10): Promise<Keyword[]> {
  const queryParams = new URLSearchParams();
  queryParams.append('sort', 'usage_count');
  queryParams.append('order', 'desc');
  if (limit) {
    queryParams.append('limit', String(limit));
  }

  const endpoint = `/api/keywords?${queryParams.toString()}`;
  return get<Keyword[]>(endpoint);
}

File: backend/app/routers/keywords.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from typing import List
import unicodedata

from app.db import get_db
from app.models.keyword import Keyword
from app.schemas.keyword import KeywordResponse, KeywordCreateRequest
from datetime import datetime

router = APIRouter(prefix="/api/keywords", tags=["keywords"])


def normalize_text(s: str) -> str:
    """
    正規化処理：
    - NFKC Normalize
    - ケースフォールディング
    """
    return unicodedata.normalize("NFKC", s).casefold()


@router.get("", response_model=List[KeywordResponse])
def list_keywords(db: Session = Depends(get_db)):
    """キーワード一覧を取得（使用回数順）"""
    return db.query(Keyword).order_by(Keyword.usage_count.desc()).all()


@router.get("/search", response_model=List[KeywordResponse])
def search_keywords(
    q: str = Query(..., min_length=1),
    db: Session = Depends(get_db),
):
    """キーワード検索（正規化して normalized_name に対して検索、使用回数順）"""
    nq = normalize_text(q)

    return (
        db.query(Keyword)
        .filter(Keyword.normalized_name.like(f"%{nq}%"))
        .order_by(Keyword.usage_count.desc())
        .all()
    )

@router.post("", response_model=KeywordResponse, status_code=201)
def create_keyword(
    request: KeywordCreateRequest,
    db: Session = Depends(get_db),
):
    """
    キーワードを作成
    既存の場合は既存のものを返す
    """
    normalized_name = normalize_text(request.name)
    
    # 既存チェック
    existing = db.query(Keyword).filter(
        Keyword.normalized_name == normalized_name
    ).first()
    
    if existing:
        return existing
    
    # 新規作成
    new_keyword = Keyword(
        name=request.name.strip(),
        normalized_name=normalized_name,
        usage_count=0,
        created_at=datetime.now()
    )
    db.add(new_keyword)
    db.commit()
    db.refresh(new_keyword)
    
    return new_keyword